avro4s-core/src/main/scala/com/sksamuel/avro4s/AvroValue.scala

```diff
@@ -0,0 +1,63 @@
+package com.sksamuel.avro4s
+
+import java.nio.ByteBuffer
+
+import org.apache.avro.generic.{GenericEnumSymbol, GenericFixed, GenericRecord}
+import org.apache.avro.util.Utf8
+
+sealed trait AvroValue
+
+object AvroValue {
+
+  import scala.collection.JavaConverters._
+
+  case class AvroString(s: String) extends AvroValue
+  case class AvroByteArray(bytes: Array[Byte]) extends AvroValue
+  case class AvroGenericFixed(fixed: GenericFixed) extends AvroValue
+  case class AvroBoolean(boolean: Boolean) extends AvroValue
+  case class AvroShort(short: Short) extends AvroValue
+  case class AvroByte(byte: Byte) extends AvroValue
+  case class AvroDouble(double: Double) extends AvroValue
+  case class AvroFloat(float: Float) extends AvroValue
+  case class AvroInt(int: Int) extends AvroValue
+  case class AvroLong(long: Long) extends AvroValue
+  case class AvroEnumSymbol(symbol: GenericEnumSymbol[_]) extends AvroValue
+  case class AvroList(list: List[AvroValue]) extends AvroValue
+  case class AvroMap(map: Map[String, AvroValue]) extends AvroValue
+  case object AvroNull extends AvroValue
+
+  case class AvroRecord(record: GenericRecord) extends AvroValue {
+    def get(name: String): Option[AvroValue] = Option(record.get(name)).map(AvroValue.unsafeFromAny)
+    def get(i: Int): Option[AvroValue] = Option(record.get(i)).map(AvroValue.unsafeFromAny)
+  }
+
+  private[avro4s] def unsafeFromAny(a: Any): AvroValue = a match {
+    case r: GenericRecord => AvroRecord(r)
+    case s: String => AvroString(s)
+    case u: Utf8 => AvroString(u.toString)
+    case c: CharSequence => AvroString(c.toString)
+    case b: Boolean => AvroBoolean(b)
+    case g: GenericFixed => AvroGenericFixed(g)
+    case b: ByteBuffer => AvroByteArray(b.array())
+    case a: Array[Byte] => AvroByteArray(a)
+    case a: Array[_] => AvroList(a.toList.map(AvroValue.unsafeFromAny))
+    case m: java.util.Map[_, _] => AvroMap(m.asScala.map { case (key, value) => key.toString -> unsafeFromAny(value) }.toMap)
+    case c: java.lang.Iterable[_] => AvroList(c.asScala.map(AvroValue.unsafeFromAny).toList)
+    case c: java.util.Collection[_] => AvroList(c.asScala.map(AvroValue.unsafeFromAny).toList)
+    case f: Float => AvroFloat(f)
+    case f: java.lang.Float => AvroFloat(f)
```

Comment:
I think the `java.lang` variants aren't really necessary here. `a` is `Any` anyway, so it's always going to be boxed.

Reply:
I definitely had an issue in the past, which is why all the decoders matched on both java.lang and scala types. Perhaps that has gone away, but leaving it in won't hurt.

Reply:
The Java types are what you get back as types from GenericRecord fields (i.e. this is what the Avro lib will produce). The Scala collection types the ones we could drop, but are there for convenience, probably.